### ISSUE TITLE
TFT Upload fails on 91%

### ISSUE THREAD

**Issue by remmartens** [Blackymas/NSPanel_HA_Blueprint#1692]
### TFT Version

4.2.4

### ESPHome Version

4.2.4

### Blueprint Version

4.2.4

### Panel Model

NSPanel EU Model

### What is the bug?

Everytime when i upload the EU Model tft it fails at 91%

### Steps to Reproduce

Select nspanel_eu.tft from the HomeAssistant GUI and press the upload button.
Get stuck at 91%
Powercyle the panel 
Select Blank.tft from the HomeAssistant GUI and press the upload button
Succesfully uploaded.
Try again eu.tft -> get stuck

Saw on other issu's try with dev repo. -> no differents
Try with local version or with the get from github -> same output -> get stuck at 91%

As soon as you get the recv_sting [] it's stuck. on the panel it is still on 91% and frozen.
Diffrent bautrate did also not help

### Your panel's YAML

```yaml
substitutions:

###### CHANGE ME START ######

  device_name: "nspanelremco" 
  wifi_ssid: !secret wifi_ssid
  wifi_password: !secret wifi_password
  

  nextion_update_url: "http://192.168.178.12:8123/local/nspanel/nspanel_eu.tft" # URL to local tft File
#  nextion_update_url: "https://raw.githubusercontent.com/Blackymas/NSPanel_HA_Blueprint/main/nspanel_us.tft" # URL to Github
  #nextion_update_url: "http://192.168.178.12:8123/local/nspanel/nspanel_blank.tft"

##### CHANGE ME END #####

##### DO NOT CHANGE ANYTHING! #####

 packages:
  remote_package:
    url: https://github.com/Blackymas/NSPanel_HA_Blueprint
    ref: dev
    files:
      - nspanel_esphome.yaml # Core package
      # Optional advanced and add-on configurations
      # - advanced/esphome/nspanel_esphome_advanced.yaml
      # - nspanel_esphome_addon_climate_cool.yaml
      # - nspanel_esphome_addon_climate_heat.yaml
      # - nspanel_esphome_addon_climate_dual.yaml
    refresh: 300s

 esp32:
  framework:
    type: esp-idf
##### DO NOT CHANGE ANYTHING! #####
```

**Comment by remmartens** [comment]
[logs_nspanel-remco_run (1).txt](https://github.com/Blackymas/NSPanel_HA_Blueprint/files/13981762/logs_nspanel-remco_run.1.txt)

**Comment by edwardtfn** [comment]
> As soon as you get the recv_sting [] it's stuck

This is a good catch. I will see if I can change the component to handle that response nicely. It shoukd at least report that the transfer failed.

> Diffrent bautrate did also not help

In fact, changing the baud rate in the UI will change the rate for normal operation only. That wouldn't change the tft upload baud rate, which always tries 921600bps first.
In order to change that, please add this to your substitutions: `upload_tft_baud_rate: "115200"`

**Comment by edwardtfn** [comment]
@remmartens, could you please add this customization to your yaml, install it into your panel and then try the transfer again?

```yaml
external_components:
  - source:
      type: git
      url: https://github.com/edwardtfn/esphome
      ref: nextion-17
    components: [ nextion ]
```

In fact this is not changing that much in the upload engine, so I would expect the same error, but with this customization I'm trying to log of the available memory and handle the `recv_sting []` nicely, so ESPHome can detect the transfer isn't happening and continue with the retry process.

Please collect logs and share it here. 😉

**Comment by edwardtfn** [comment]
I just found a point to optimize memory on that library, so I think we have increased a bit the chances for a successful transfer. although I'm not sure that issue in your log was related to memory, so please send your logs anyways.

**Comment by remmartens** [comment]
Dear @edwardtfn thank you for the reply. 
I tried to add code you mentioned. It's stopping the upload at the correct time. But is stuck afterwards. so powercycle is needed. 
See the logs for the output. 
[logs_nspanel-remco_run (2).txt](https://github.com/Blackymas/NSPanel_HA_Blueprint/files/14001436/logs_nspanel-remco_run.2.txt)

**Comment by remmartens** [comment]
Also when i changes the bautrate to 115200 it's giving the same behavior. Can it be that the file is to big for the display? Because uploading the blank one is always succesfull. Can i do something to clean the nextion space? 

> > As soon as you get the recv_sting [] it's stuck
> 
> This is a good catch. I will see if I can change the component to handle that response nicely. It shoukd at least report that the transfer failed.
> 
> > Diffrent bautrate did also not help
> 
> In fact, changing the baud rate in the UI will change the rate for normal operation only. That wouldn't change the tft upload baud rate, which always tries 921600bps first. In order to change that, please add this to your substitutions: `upload_tft_baud_rate: "115200"`

**Comment by edwardtfn** [comment]
I don't think there's a way to free-up Nextion memory that will be different than installing a new TFT.
It could be that for some reason the memory is limited, but then it should have failed with previous versions, as the TFT haven't changed that much.

One thing you can try is to install an early version of the TFT just to try.
In Home Assistant, go to Developer Tools, select the tab "Services", type "tft" in the service name area and you will see a drop down where your panel's service to upload TFT will be listed. Select there.
On the server setting, add the url for your TFT file, then click "Call service".
The url is like this: `https://raw.githubusercontent.com/Blackymas/NSPanel_HA_Blueprint/v4.2.4/nspanel_eu.tft`
You can replace `v4.2.4` by another version you want to install. Let's say you want to install v4.1.4, your url will be `https://raw.githubusercontent.com/Blackymas/NSPanel_HA_Blueprint/v4.1.4/nspanel_eu.tft`

**Comment by edwardtfn** [comment]
Another thing you could try is installing another file, like `https://raw.githubusercontent.com/Blackymas/NSPanel_HA_Blueprint/v4.2.4/nspanel_us_land.tft`, again, just to double check if the error is in another place.

**Comment by filgr8** [comment]
When I was flashing tft file on my 2 nspanel it would fail at 91% as well, but I would upload again and it would finish uploading very quickly.

**Comment by edwardtfn** [comment]
> But is stuck afterwards. so powercycle is needed.

I've added a button to power cycle the display, so hopefully you don't need to power cycle the entire panel. This is in `dev` and will be included in the next release.

**Comment by edwardtfn** [comment]
@remmartens, if you are still having issues, could you please share a new log of your upload process?

**Comment by remmartens** [comment]
Dear @edwardtfn Yes it's still not fixed. I will try to test your new update later this week. I'm not be able to test it earlier because of work. When i have an update i will add this to this issu

**Comment by remmartens** [comment]
[logs_nspanel-remco_validate.txt](https://github.com/Blackymas/NSPanel_HA_Blueprint/files/14102732/logs_nspanel-remco_validate.txt)
Also an log for the validate. -> needed to remove the section as mentioned here because this can't compile also https://github.com/Blackymas/NSPanel_HA_Blueprint/issues/1692#issuecomment-1902466422 
If you need more logs let me know so i can test it for you.

**Comment by edwardtfn** [comment]
Please try cleaning your build files and try compiling again. I've just tried with no issues.

https://github.com/Blackymas/NSPanel_HA_Blueprint/blob/main/docs/error_compiling.md#clean-up-build-files

**Comment by remmartens** [comment]
Dear @edwardtfn, tried to update and indeed after clean it works to update esphome. Tried again to update tft EU and tried US without succes. Stuck on 91%. Reboot is to short so i had to powercycle it. (Blinking white screen with message in esphome “nextion not connected” ) after powercycle the System date error is back and nextion connects to esphome. 

Will try later this week again so i can send you logs. Didnt manage to save them when i was working with it.

**Comment by edwardtfn** [comment]
Could you please share the logs of the upload this time? I've added more info about the memory consumption in the logs.

**Comment by remmartens** [comment]
[logs_nspanel-remco_logs.txt](https://github.com/Blackymas/NSPanel_HA_Blueprint/files/14184540/logs_nspanel-remco_logs.txt)
Dear Edwardtfn see the attached logs

**Comment by remmartens** [comment]
[logs_nspanel-remco_run (1).txt](https://github.com/Blackymas/NSPanel_HA_Blueprint/files/14184623/logs_nspanel-remco_run.1.txt)
Dear edwardtfn see the attached logs.

**Comment by edwardtfn** [comment]
Sorry, I've just noticed you are running a `dev` version, which is Ok, but then In this case I would ask you to **remove** that `external_components` area from your yaml:

Remove this:
```yaml
external_components:
  - source:
      type: git
      url: https://github.com/edwardtfn/esphome
      ref: nextion-17
    components: [ nextion ]
    refresh: 1s
```

After removing that, please flash it again (wirelessly is fine), then try the upload again.

Please don't forget to share logs if it still failiing.

**Comment by edwardtfn** [comment]
We just released a patch with solves this: [v4.2.5 - Celebrating 1000 Stars with Enhanced Functionality and UI Improvements](https://github.com/Blackymas/NSPanel_HA_Blueprint/releases/tag/v4.2.5)
Could you please update your system and let me know if the issue was solved?
